### DISCUSSION TITLE
Have Record implement Decodable by default

### DISCUSSION THREAD

**Discussion by oddanderson** [groue/GRDB.swift#1146] [Ideas]
In the guide for querying using Joins, the guide relies heavily on structs in favor of objects. When it comes to the filter/querying guide, this makes the examples really simple because struct implements decodable already, assuming you only have a limited set of property types.

However, if one of your object types in a struct is a Record, you need to implement Decodable's`init(from: decoder)` in both the struct AND the record object, such that I end up with:

```
struct EventInfo: Decodable, FetchableRecord {
    struct PartialContact: Decodable, FetchableRecord {
        var contactName: String
        var contactId: Int64
    }
    var event: Event
    var contact: PartialContact
    
    init(from decoder: Decoder) throws {
        event = try Event.init(from: decoder)
        contact = try PartialContact.init(from: decoder)
    }
}
class Event: Record, Decodable {
    
    private enum CodingKeys: String, CodingKey {
        case id
        ....
    }
    required init(from decoder:Decoder) throws {
        super.init()
        let values = try decoder.container(keyedBy: CodingKeys.self)
        id = try values.decode(Int64?.self, forKey: .id)
        ...
    }
```

As I was doing so, it seemed like I was writing a fair amount of duplicate code relative to the other functions I need to implement for Record, specifically `enum Columns: String, ColumnExpression`, `required init(row: Row)`, and `override func encode(to container: inout PersistenceContainer)` 
 
Curious if it makes sense to use those definitions to enable Record to implement Decodable for free

**Comment by groue** [comment]
> Curious if it makes sense to use those definitions to enable Record to implement Decodable for free

It is not possible to have `Record` conform to `Decodable` without forcing *all* subclasses to conform to Decodable as well (even for people who don't care about `Decodable`). You can try in a fresh playground:

```swift
class Record: Decodable { }

// Compiler error: 'required' initializer 'init(from:)' must be provided by subclass of 'Record'
class MyRecord: Record {
    var id: Int64

    init(id: Int64) {
        self.id = id
    }
}
```
